controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts

```diff
@@ -0,0 +1,364 @@
+      step = 15;
+      break;
+    }
+    case 24: {
+      step = 5;
+      break;
+    }
+    case 1: {
+      step = 60;
+      break;
+    }
+  }
+
+  const x = (60 / step) * rangeInHours;
+
+  for (let i = 0; i <= x; i++) {
+    range.unshift({
+      timestamp: String(startTimestamp - i * step * 60 * 1000),
+    });
+  }
+
+  return range;
+};
+
+interface GetRangeQueryPropsOptions {
+  /**
+   * Range in hours
+   */
+  range: number;
+  /**
+   * End date in milliseconds
+   */
+  endDate: number;
+}
+
+interface MetricsParams {
+  organizationId: string;
+  graphId: string;
+  graphName: string;
+}
+
+interface GetMetricsProps {
+  range?: number;
+  endDate?: number;
+  params: MetricsParams;
+}
+
+export class MetricsDashboardRepository {
+  constructor(private client: PrometheusClient) {}
+
+  /**
+   * Get request rate metrics
+   */
+  public async getRequestRateMetrics({ range = 24, endDate = getUtcUnixTime(), params }: GetMetricsProps) {
+    const prange = '15m';
+
+    // get median of requests in last [range] hours
+    const median = await this.client.query({
+      query: `quantile(0.5, sum(rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}}[${range}h])) * 60)`,
+    });
+
+    const prevMedian = await this.client.query({
+      query: `quantile(0.5, sum(rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}}[${range}h] offset ${range}h)) * 60)`,
+    });
+
+    // get top 5 operations in last [range] hours
+    const top5 = await this.client.query({
+      query: `topk(5, sum by (wg_operation_name) (rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}}[${range}h])) * 60)`,
+    });
+
+    // get requests in last [range] hours in series of [step]
+    const result = await this.client.queryRange({
+      query: `sum (rate(cosmo_router_http_requests_total{${this.getQueryLabels(params)}}[${prange}]) * 60)`,
+      ...this.getRangeQueryProps({
+        range,
+        endDate,
+      }),
+    });
+
+    const prevResult = await this.client.queryRange({
+      query: `sum (rate(cosmo_router_http_requests_total{${this.getQueryLabels(params)}}[${prange}]) * 60)`,
+      ...this.getRangeQueryProps({
+        range,
+        endDate: endDate - range * 60 * 60 * 1000,
+      }),
+    });
+
+    return {
+      value: median.data.result[0]?.value[1],
+      previousValue: prevMedian.data.result[0]?.value[1],
+      top: top5.data.result.map((v: any) => ({
+        name: v.metric.wg_operation_name,
+        value: v.value[1],
+      })),
+      series: this.mapSeries(endDate, range, result.data.result[0].values, prevResult.data.result[0]?.values),
+    };
+  }
+
+  /**
+   * Get latency metrics
+   */
+  public async getLatencyMetrics({ range = 24, endDate = getUtcUnixTime(), params }: GetMetricsProps) {
+    const prange = '15m';
+
+    // get p95 of requests in last [range] hours
+    const latencyP95 = await this.client.query({
```

Comment:
I think you missed handling error cases. I'd ensure that the client `query` always returns a valid result when it doesn't throw an error. You need to check for errors props in the response.

See https://prometheus.io/docs/prometheus/latest/querying/api/

Reply:
I've updated the error handling, I'm returning all errors now from the repository methods, should we log them, pass them back to the client?